common/src/main/java/io/netty/util/internal/RefCnt.java

```diff
@@ -0,0 +1,463 @@
+/*
+ * Copyright 2025 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal;
+
+import io.netty.util.IllegalReferenceCountException;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import static io.netty.util.internal.ObjectUtil.checkPositive;
+
+/**
+ * Concrete class for reference counting implementations.
+ * Provides a factory method to create instances using the most efficient available atomic updater.
+ */
+public class RefCnt {
+
+    private static final int UNSAFE = 0;
+    private static final int VAR_HANDLE = 1;
+    private static final int ATOMIC_UPDATER = 2;
+    private static final int REF_CNT_IMPL;
+
+    static {
+        if (PlatformDependent.hasUnsafe()) {
+            REF_CNT_IMPL = UNSAFE;
+        } else if (PlatformDependent.hasVarHandle()) {
+            REF_CNT_IMPL = VAR_HANDLE;
+        } else {
+            REF_CNT_IMPL = ATOMIC_UPDATER;
+        }
+    }
+
+    /*
+     * Implementation notes:
+     *
+     * For the updated int field:
+     *   Even => "real" refcount is (refCnt >>> 1)
+     *   Odd  => "real" refcount is 0
+     */
+    volatile int value;
+
+    public RefCnt() {
+        switch (REF_CNT_IMPL) {
+        case UNSAFE:
+            UnsafeRefCnt.init(this);
+            break;
+        case VAR_HANDLE:
+            VarHandleRefCnt.init(this);
+            break;
+        case ATOMIC_UPDATER:
+        default:
+            AtomicRefCnt.init(this);
+            break;
+        }
+    }
+
+    private static long getUnsafeOffset(Class<?> clz, String fieldName) {
+        try {
+            if (PlatformDependent.hasUnsafe()) {
```

Comment:
There is no need for this, in theory

Reply:
I have not enabled VarHandle in native image, but if you want - give it a shot or we can do it in a follow up PR too